### PR TITLE
Adding support for mixed precision quantization configs.

### PR DESCRIPTION
**Pull request by singh-mitali**
Using mixed precision config improves inference accuracy and throughput significantly over int8.

### CODE REVIEW

**Comment by mitalisi** [comment]
> Looks good to me. I see some numbers pasted in b/351878551, are those the end-2-end runs? They are lower than the int8 weight but still better than sax numbers.

Yes those are end-to-end jetstream benchmarking runs. Numbers are better for mixed_precision than int8w. Will discuss offline to make sure we are looking at the same numbers.

File: MaxText/configs/quantization/mp_scale.json [hunk 1]
```diff
@@ -0,0 +1,8 @@
+{
```
Comment (vipannalla) [inline MaxText/configs/quantization/mp_scale.json:1]: Add doc book or reference to AQT example/code.

File: MaxText/configs/base.yml [hunk 2]
```diff
@@ -59,7 +59,14 @@ save_config_to_gcs: False
 
 # Activation dtypes.
 dtype: "bfloat16"
-quantization: "" #defaults to no quantization, i.e. bf16. possible alternative setting is 'int8' or use fp8 to run with 8-bit floating-point GeMMs on NVIDIA GPUs.
+# defaults to no quantization, i.e. bf16. Possible alternative settings are as follows:
+# 'int8' for dynamic range quantization using 8-bits
```
[the remark this answers is not in the document]
Comment (singh-mitali) [reply MaxText/configs/base.yml:64]: I have updated the comments for more clarity.

File: MaxText/configs/quantization/mp_subchannel.json [hunk 3]
```diff
@@ -0,0 +1,8 @@
+{
```
[the remark this answers is not in the document]
Comment (singh-mitali) [reply MaxText/configs/quantization/mp_subchannel.json:1]: It getting loaded here:     https://github.com/google/maxtext/blob/74c9d6bb01e5437a340c36131a420255b6a7aae9/MaxText/layers/quantizations.py#L200

File: MaxText/configs/quantization/mp_scale.json (hunk 1 above)
Comment (singh-mitali) [reply MaxText/configs/quantization/mp_scale.json:1]: Added more comments in config file. Also an example of how to use it is here: https://github.com/google/maxtext/blob/74c9d6bb01e5437a340c36131a420255b6a7aae9/end_to_end/tpu/test_decode_save_quantized_ckpt.sh#L69

File: MaxText/layers/quantizations.py [hunk 4]
```diff
@@ -104,37 +167,67 @@ def dot_general_cls(self, mesh_axes: Tuple[str, ...] = ()):
...
+
+def _get_mixed_precision_quant_config(config, config_file):
+  """Set quantization params based on user configuration."""
+  with open(config_file, "r") as infile:
+    mixed_precision_config = json.load(infile)
+  ret_config = {}
+  ret_config["default"] = [aqt_config.dot_general_make(lhs_bits=None, rhs_bits=8), -1]
+  for layer_name_re, layer_quantization_config in mixed_precision_config.items():
+    rhs_num_bits = layer_quantization_config.get("bits", 8)
+    tile_size = layer_quantization_config.get("tile_size", -1)
```
Comment (gobbleturk) [inline MaxText/layers/quantizations.py:205]: can we include documentation in the base.yml of the defaults and the behavior of the defaults - that tile_size=-1 signifies "don't tile"

**Review by gobbleturk — APPROVED** [review]

File: MaxText/configs/quantization/mp_scale.json
[the remark this answers is not in the document]
Comment (singh-mitali) [reply MaxText/configs/quantization/mp_scale.json:1]: its a json data file - will not accept comments.

File: MaxText/configs/quantization/mp_subchannel.json
[the remark this answers is not in the document]
Comment (singh-mitali) [reply MaxText/configs/quantization/mp_subchannel.json:1]: Again a json data file.

File: MaxText/configs/quantization/mp_scale.json
[the remark this answers is not in the document]
Comment (singh-mitali) [reply MaxText/configs/quantization/mp_scale.json:1]: Done, Currently we have tuned them for llama models based on experimentation.

File: MaxText/configs/quantization/mp_subchannel.json
[the remark this answers is not in the document]
Comment (singh-mitali) [reply MaxText/configs/quantization/mp_subchannel.json:1]: Added comments. This is an alternate mixed-precision config for subchannel quantization.

File: MaxText/configs/base.yml [hunk 5]
```diff
@@ -59,7 +59,14 @@ save_config_to_gcs: False
...
+# 'int8w' for weights only quantization using 8-bits
+# 'int4w' for weights only quantization using 4-bits
+# 'intmp' for mixed precision quantization based on config file
+# 'fp8' for 8-bit floating-point GeMMs on NVIDIA GPUs.
+quantization: ""
+quant_cfg_path: "" # Path to file with quantization config - only used for mixed precision.
```
[the remark this answers is not in the document]
Comment (singh-mitali) [reply MaxText/configs/base.yml]: Done

File: MaxText/layers/quantizations.py (hunk 4 above)
Comment (singh-mitali) [reply MaxText/layers/quantizations.py:205]: Done

File: end_to_end/tpu/test_decode_load_quantized_ckpt.sh [hunk 6]
```diff
@@ -1,13 +1,19 @@
 #!/bin/sh
 
+# # Example run: bash end_to_end/tpu/test_decode_load_quantized_ckpt.sh  -m llama2-70b -r test -s decode -n
```
[the remark this answers is not in the document]
Comment (singh-mitali) [reply end_to_end/tpu/test_decode_load_quantized_ckpt.sh:3]: Removed

**Pull request merged**